### PR TITLE
[KYUUBI #3180] Add component version util

### PR DESCRIPTION
**Pull request by deadwind4**
### _Why are the changes needed?_

Add component version util class

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: extensions/spark/kyuubi-spark-authz/pom.xml [hunk 1]
```diff
@@ -39,6 +39,11 @@
     </properties>
 
     <dependencies>
+        <dependency>
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/pom.xml:42]: we can duplicate the code, authz aims to be independent

File: kyuubi-common/src/test/scala/org/apache/kyuubi/engine/ComponentVersionSuite.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    assert(version.majorVersion === 1)
+    assert(version.minorVersion === 12)
+    assert(version.patchVersion === "4")
+  }
+
+  test("parse snapshot version") {
+    val version = new ComponentVersion("2.14.8-SNAPSHOT")
+    assert(version.majorVersion === 2)
+    assert(version.minorVersion === 14)
+    assert(version.patchVersion === "8-SNAPSHOT")
```
Comment (pan3793) [inline kyuubi-common/src/test/scala/org/apache/kyuubi/engine/ComponentVersionSuite.scala:35]: I don't think this is right, patch version should be "8".

In practice, when we fork and maintain a internal version of an open-source project, we add additional suffix in version, e.g. the upstream version is `1.2.3`, the internel version may be `1.2.3.1`, `1.2.3.20220781`, `1.2.3-org`, and those version should be compatible w/ `1.2.3`

Comment (deadwind4) [reply kyuubi-common/src/test/scala/org/apache/kyuubi/engine/ComponentVersionSuite.scala:35]: The patch version is in various forms. I delete the patch version in this class and think about it in the future.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/ComponentVersion.scala [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.engine
+
+/**
+ * Encapsulate a component (Kyuubi/Spark/Hive/Flink) version for the convenience of version checks.
+ */
+class ComponentVersion(versionString: String) {
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/engine/ComponentVersion.scala:23]: case class TernaryVersion(major: Int, minor: Int, patch: Int)

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/ComponentVersion.scala [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        (m.group(1).toInt, m.group(2).toInt, m.group(3))
+      case None =>
+        throw new IllegalArgumentException(s"Tried to parse '$versionString' as a project" +
+          s" version string, but it could not find the major, minor and patch version numbers.")
+    }
+}
+
+object ComponentVersion {
+
+  def isVersionAtMost(targetVersionString: String, runtimeVersionString: String): Boolean = {
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/engine/ComponentVersion.scala:37]: define `def isVersionAtMost(targetVersionString: String)` in the companion class

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/ComponentVersion.scala [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      targetVersionString,
+      runtimeVersionString,
+      (targetMajor: Int, targetMinor: Int, runtimeMajor: Int, runtimeMinor: Int) => {
+        (runtimeMajor > targetMajor) || {
+          runtimeMajor == targetMajor && runtimeMinor >= targetMinor
+        }
+      })
+  }
+
+  def isVersionEqualTo(targetVersionString: String, runtimeVersionString: String): Boolean = {
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/engine/ComponentVersion.scala:59]: ditto

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/util/ComponentVersion.scala [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.plugin.spark.authz.util
+
+/**
+ * Encapsulate a component (Kyuubi/Spark/Hive/Flink) version for the convenience of version checks.
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/util/ComponentVersion.scala:21]: this is a spark oriented module, should not have opportunities to work with hive flink

**Comment by yaooqinn** [comment]
there are also some duplicated functions in Utils.class

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/ComponentVersion.scala (hunk 3 above)
Comment (deadwind4) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/engine/ComponentVersion.scala:23]: I will try to implement `case class BinaryVersion(major: Int, minor: Int)`. IMO, The patch version is not considered at present.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/util/ComponentVersion.scala (hunk 6 above)
Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/util/ComponentVersion.scala:21]: we can leave this module alone in this PR.

Comment (deadwind4) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/util/ComponentVersion.scala:21]: Should I revert this module in this PR?

Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/util/ComponentVersion.scala:21]: it's up to you, i am ok with both

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/ComponentVersion.scala (hunk 3 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/engine/ComponentVersion.scala]: How about naming it `SemanticVersion`?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/ComponentVersion.scala (hunk 4 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/engine/ComponentVersion.scala:60]: Agreed. 

The following API is more intuitive.
```
ComponentVersion("3.2.2").isVersionAtMost("3.3")
```

File: extensions/spark/kyuubi-spark-connector-common/src/main/scala/org/apache/kyuubi/spark/connector/common/SparkUtils.scala [hunk 7]
```diff
@@ -21,58 +21,15 @@ import org.apache.spark.SPARK_VERSION
 
 object SparkUtils {
```
Comment (deadwind4) [inline extensions/spark/kyuubi-spark-connector-common/src/main/scala/org/apache/kyuubi/spark/connector/common/SparkUtils.scala:22]: Should I delete this `SparkUtils` class?

**Review by pan3793 — APPROVED** [review]

File: extensions/spark/kyuubi-spark-connector-common/src/main/scala/org/apache/kyuubi/spark/connector/common/SparkUtils.scala (hunk 7 above)
Comment (pan3793) [reply extensions/spark/kyuubi-spark-connector-common/src/main/scala/org/apache/kyuubi/spark/connector/common/SparkUtils.scala:22]: I'm neutral, either is fine.

**Comment by pan3793** [comment]
Thanks, merging to master

**Pull request closed without merging**